### DISCUSSION TITLE
Classist vs Mockist

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#235] [Spring Test]
.

**Comment by Kimprodp** [comment]
## Classist
Classist 테스트 스타일은 주로 통합 테스트와 유사하며, 객체들을 모킹(mocking)하지 않고 실제 객체들을 사용하여 테스트를 진행하는 방식이다.

특징
- 객체 간의 실제 상호작용을 테스트하기 위해 테스트에 필요한 협력 객체들을 그대로 사용한다.
- 행위가 수행된 다음 상태를 확인하게 된다. (상태 검증)

장점
- 시스템의 실제 동작을 더 잘 반영한다.
- 객체 간의 상호작용을 정확하게 테스트할 수 있다.

단점
- 복잡한 검증일 경우 상태 검증을 위해 검증될 데이터를 만들기 쉽지 않다.
- 설정이 복잡하고 테스트가 느릴 수 있다.
- 서비스에 장애가 있을 경우 테스트가 격리되지 않고 내부 요소들까지 장애가 발생하면서 테스트가 실패한다.

## Mockist
Mockist 테스트 스타일은 단위 테스트에 초점을 맞추며, 객체 간의 협력을 테스트하기 위해 목 객체(mock objects)를 사용하는 방식이다.

특징
- 테스트 대상 객체가 다른 객체와 어떻게 상호작용 하는 지를 테스트 한다.
- 목 객체를 사용하여 객체 간의 상호작용을 테스트한다.
- 인터페이스를 통해 협력 객체를 모킹한다.
- 가짜 객체를 사용하기 때문에 내부 구현을 커스텀한 응답으로 응답해야 한다.
- 특정 행위가 실행되었는지 검증한다. (행위 검증)

장점
- 행위만 검증하기 때문에 검증 데이터를 만들 필요가 없다.
- 테스트가 빠르다.
- 테스트가 독립적이기 때문에 변경이 다른 부분에 영향을 미치지 않는다.

단점
- 실제 동작을 잘 반영하지 못할 수 있다.
- 목 객체의 설정이 복잡할 수 있다.
- Classist와 달리 내부적으로 장애가 발생해도 단순히 버그가 발생한 클래스에 대해서만 테스트 실패를 하게 된다.

**Comment by seunghye218** [comment]
## 단위 테스트에서 목과 스텁을 사용하는 것에 관해 두 가지 의견

- Mockist(목찬성론자 or 런던 학파)
    - 개발자는 단위 테스트 코드에서 의존성을 실제로 사용하는 것을 피해야 하고 대신 목을 사용해야 한다고 주장
    - 실제 의존성의 사용을 피하고 목을 사용하는 것은 의존성 코드로부터 값을 받는 경우 스텁을 사용한다는 것을 의미
- Classicist(고전주의자 or 디트로이트 학파)
    - 목과 스텁은 최소한으로 사용되어야 하고 개발자는 테스트에서 의존성을 실제로 사용하는 것을 최우선으로 해야 한다고 주장
    - 실제 의존성을 사용하는 것이 가능하지 않을 때, 페이크를 사용하는 것을 선호
    - 목과 스텁은 실제 의존성이나 페이크를 사용하는 것이 불가능할 때만 최후의 수단으로 사용

주요 차이점 중 하나는 목을 사용한 테스트는 상호작용을 테스트하는 반면, 고전주의 방법을 사용한 테스트는 코드의 결과 상태와 의존성을 테스트하는 경향이 있습니다.

### 테스트 대상:
- 목을 사용한 테스트: 상호 작용을 테스트합니다. 즉, 시험 대상 코드가 다른 객체와 어떻게 상호 작용하는지 확인합니다.
- 고전주의 방법을 사용한 테스트: 코드의 결과 상태와 의존성을 테스트합니다. 즉, 시험 대상 코드가 실행된 후 어떤 결과를 만들어내는지, 그리고 코드가 어떤 의존성을 가지고 있는지 확인합니다.

### 검증 방식:
- 목 접근법: 시험 대상 코드가 어떻게 "행동"하는지 확인합니다. 즉, 코드가 예상대로 메서드를 호출하고, 올바른 인수를 전달하고, 예상되는 결과를 반환하는지 확인합니다.
- 고전주의 접근법: 코드를 실행하는 최종 결과가 무엇인지 확인합니다. 즉, 테스트 케이스가 성공적으로 실행되었는지, 시험 대상 코드가 예상대로 동작했는지 확인합니다. 이 결과는 어떻게 달성되는지는 크게 중요하지 않습니다.

### 장점:
- 목 사용을 지지하는 주장:

1. 단위 테스트가 더욱 격리됩니다. 목을 사용하면 테스트가 의존성에 대한 것들을 테스트하지 않는다는 것을 의미합니다. 즉, 특정 코드에 문제가 있을 때 해당 코드에 대한 단위 테스트에서만 테스트 실패를 유발하며, 이 코드에 의존하는 다른 코드에 대한 테스트는 실패하지 않습니다.
2. 테스트 코드 작성 및 유지 관리가 더욱 용이합니다. 목을 사용하면 테스트 코드를 더욱 간결하고 명확하게 작성할 수 있습니다. 또한, 의존성 코드의 변경에 따른 테스트 코드의 변경 작업도 줄어듭니다.
3. 테스트 실행 속도가 향상됩니다. 목을 사용하면 실제 의존성 객체를 대신하기 때문에 테스트 실행 속도가 향상될 수 있습니다.

- 고전주의 접근법의 장점:
1. 실제 시스템 동작을 더 잘 반영합니다. 고전주의 접근법은 실제 의존성 객체를 사용하기 때문에 테스트 시나리오가 실제 시스템 동작을 더 잘 반영할 수 있습니다.
2. 테스트 코드 작성 및 유지 관리가 간편할 수 있습니다. 특히, 의존성 객체와 시험 대상 코드가 밀접하게 결합되어 있는 경우, 고전주의 접근법이 더욱 간편하게 테스트 코드를 작성하고 유지 관리할 수 있습니다.

### 결론

두 가지 방법 모두 장단점이 있으며, 어떤 방법을 사용할지는 상황에 따라 달라집니다. 일반적으로, 다음과 같은 경우 목을 사용하는 것이 좋습니다.
- 시험 대상 코드가 복잡한 의존성을 가지고 있는 경우
- 테스트 코드 작성 및 유지 관리의 용이성이 중요한 경우
- 테스트 실행 속도가 중요한 경우

반면에, 다음과 같은 경우 고전주의 접근법을 사용하는 것이 좋습니다.
- 실제 시스템 동작을 정확하게 반영해야 하는 경우
- 의존성 객체와 시험 대상 코드가 밀접하게 결합되어 있는 경우
